packages/firestore/src/core/sync_engine.ts

```diff
@@ -764,55 +740,51 @@ export class SyncEngine implements RemoteSyncer, SharedClientStateSyncer {
-            // didn't lose any good docs that had been past the limit.
-            return this.localStore.executeQuery(queryView.query).then(docs => {
-              return queryView.view.computeDocChanges(docs, viewDocChanges);
-            });
-          })
-          .then((viewDocChanges: ViewDocumentChanges) => {
-            const targetChange =
-              remoteEvent && remoteEvent.targetChanges[queryView.targetId];
-            const viewChange = queryView.view.applyChanges(
-              viewDocChanges,
-              /* updateLimboDocuments= */ this.isPrimary === true,
-              targetChange
-            );
-            this.updateTrackedLimbos(
+      const processor = async (): Promise<void> => {
+        let viewDocChanges = queryView.view.computeDocChanges(changes);
+        if (viewDocChanges.needsRefill) {
+          // The query has a limit and some docs were removed, so we need
+          // to re-run the query against the local store to make sure we
+          // didn't lose any good docs that had been past the limit.
+          const docs = await this.localStore.executeQuery(queryView.query);
+          viewDocChanges = queryView.view.computeDocChanges(
+            docs,
+            viewDocChanges
+          );
+        }
+        const targetChange =
+          remoteEvent && remoteEvent.targetChanges[queryView.targetId];
+        const viewChange = queryView.view.applyChanges(
+          viewDocChanges,
+          /* updateLimboDocuments= */ this.isPrimary === true,
+          targetChange
+        );
+        this.updateTrackedLimbos(queryView.targetId, viewChange.limboChanges);
+        if (viewChange.snapshot) {
+          if (this.isPrimary) {
+            this.sharedClientState.updateQueryState(
               queryView.targetId,
-              viewChange.limboChanges
+              viewChange.snapshot.fromCache ? 'not-current' : 'current'
             );
-            if (viewChange.snapshot) {
-              if (this.isPrimary) {
-                this.sharedClientState.updateQueryState(
-                  queryView.targetId,
-                  viewChange.snapshot.fromCache ? 'not-current' : 'current'
-                );
-              }
-
-              newSnaps.push(viewChange.snapshot);
-              const docChanges = LocalViewChanges.fromSnapshot(
-                queryView.targetId,
-                viewChange.snapshot
-              );
-              docChangesInAllViews.push(docChanges);
-            }
-          })
-      );
+          }
+
+          newSnaps.push(viewChange.snapshot);
+          const docChanges = LocalViewChanges.fromSnapshot(
+            queryView.targetId,
+            viewChange.snapshot
+          );
+          docChangesInAllViews.push(docChanges);
+        }
+      };
+      queriesProcessed.push(processor);
     });
 
-    await Promise.all(queriesProcessed);
+    for (const queryProcess of queriesProcessed) {
+      await queryProcess();
+    }
```

Comment:
I think this change makes it so we process the queries serially (we wait for each promise before processing the next query) instead of in parallel (we start processing all the queries and then wait for them all to complete), which will reduce performance.

I think reverting this to how it was would probably be better.